src/sentry/auth/staff.py

```diff
@@ -0,0 +1,319 @@
+
+COOKIE_NAME = getattr(settings, "STAFF_COOKIE_NAME", "stf")
+
+COOKIE_SALT = getattr(settings, "STAFF_COOKIE_SALT", "")
+
+COOKIE_SECURE = getattr(settings, "STAFF_COOKIE_SECURE", settings.SESSION_COOKIE_SECURE)
+
+COOKIE_DOMAIN = getattr(settings, "STAFF_COOKIE_DOMAIN", settings.SESSION_COOKIE_DOMAIN)
+
+COOKIE_PATH = getattr(settings, "STAFF_COOKIE_PATH", settings.SESSION_COOKIE_PATH)
+
+COOKIE_HTTPONLY = getattr(settings, "STAFF_COOKIE_HTTPONLY", True)
+
+# the maximum time the session can stay alive
+MAX_STAFF_SESSION_AGE = timedelta(hours=4)
+
+IDLE_MAX_STAFF_SESSION_AGE = timedelta(minutes=15)
+
+ALLOWED_IPS = frozenset(getattr(settings, "STAFF_ALLOWED_IPS", settings.INTERNAL_IPS) or ())
+
+ORG_ID = getattr(settings, "STAFF_ORG_ID", None)
+
+DISABLE_SSO_CHECK_FOR_LOCAL_DEV = getattr(settings, "DISABLE_SSO_CHECK_FOR_LOCAL_DEV", False)
+
+UNSET = object()
+
+
+def is_active_staff(request: Request) -> bool:
+    if is_system_auth(getattr(request, "auth", None)):
+        return True
+    stf = getattr(request, "staff", None) or Staff(request)
+    return stf.is_active
+
+
+class Staff(ElevatedMode):
+    allowed_ips = frozenset(ipaddress.ip_network(str(v), strict=False) for v in ALLOWED_IPS)
+
+    def __init__(self, request, allowed_ips=UNSET, current_datetime=None):
+        self.uid: str | None = None
+        self.request = request
+        if allowed_ips is not UNSET:
+            self.allowed_ips = frozenset(
+                ipaddress.ip_network(str(v), strict=False) for v in allowed_ips or ()
+            )
+        self._populate(current_datetime=current_datetime)
+
+    @property
+    def is_active(self):
+        # We have a wsgi request with no user.
+        if not hasattr(self.request, "user"):
+            return False
+        # if we've been logged out
+        if not self.request.user.is_authenticated:
+            return False
+        # if staff status was changed
+        if not self.request.user.is_staff:
+            return False
+        # if the user has changed
+        if str(self.request.user.id) != self.uid:
+            return False
+        return self._is_active
+
+    def is_privileged_request(self):
+        """
+        Returns ``(bool is_privileged, str reason)``
+        """
+        allowed_ips = self.allowed_ips
+
+        # _admin should have always completed SSO to gain status.
+        # We expect ORG_ID to always be set in production.
+        if ORG_ID and not has_completed_sso(self.request, ORG_ID):
+            if not DISABLE_SSO_CHECK_FOR_LOCAL_DEV:
+                return False, "incomplete-sso"
+
+        # if there's no IPs configured, we allow assume its the same as *
+        if not allowed_ips:
+            return True, None
+        ip = ipaddress.ip_address(str(self.request.META["REMOTE_ADDR"]))
+        if not any(ip in addr for addr in allowed_ips):
+            return False, "invalid-ip"
+        return True, None
+
+    def get_session_data(self, current_datetime=None):
```

Comment:
There is a lot of checks and calculations happening in this function. Can we break those apart? I think the code would be much easier to read, and the flow/logic much more feasible to follow in the mind!

We also seem to be populating some keys and not others if a condition fails. For example, if `idl` is less than current, we will return early and not populate the rest of the keys (`exp`). Is that what we expect? Should the rest of the keys still be defaulted so they exist? Could we pull the populating/checking of those keys out to another function so that we could handle that decision more easily?